prow/cmd/crier/main.go

```diff
@@ -127,20 +129,22 @@ func (o *options) parseArgs(fs *flag.FlagSet, args []string) error {
 	fs.StringVar(&o.jobConfigPath, "job-config-path", "", "Path to prow job configs.")
 
 	// TODO(krzyzacy): implement dryrun for gerrit/pubsub
-	fs.BoolVar(&o.dryrun, "dry-run", false, "Run in dry-run mode, not doing actual report (effective for github only)")
+	fs.BoolVar(&o.dryrun, "dry-run", false, "Run in dry-run mode, not doing actual report (effective for github and Slack only)")
 
 	o.github.AddFlags(fs)
 	o.client.AddFlags(fs)
 
 	fs.Parse(args)
 
+	if len(o.gerritProjects) == 0 {
+		o.gerritProjects = nil
```

Comment:
Well-formed zero values for the struct should function here instead of needing `nil`. Why do we need this?

Reply:
A lot of the reasoning here is going to come from my OO experience, so bear with me.

It seemed _marginally_ better for an unused, uninstantiated object to be left "nil" instead of empty, for two reasons:

1) A script constructing an options object (such as the Unit tests) wouldn't need to instantiate empty sub-objects to leave it in a "correct" state.
2) Though unlikely, the behavior of the gerritclient.ProjectsFlag struct may change such that an empty object is no longer valid. The GitHubOptions object is like this: an empty instance of that struct is invalid. There are default values stored within it that are only set once "AddFlags" is called.

I'm somewhat new to Go, so I suspect there's something I'm missing here. Why is an empty struct preferable to a nil value?

Reply:
I think it's just the Go idiom -- all you say is true, just not idiomatic for Go [link](https://golang.org/doc/effective_go.html#allocation_new):

> Since the memory returned by new is zeroed, it's helpful to arrange when designing your data structures that the zero value of each type can be used without further initialization. This means a user of the data structure can create one with new and get right to work.

Reply:
It's worth noting that in this particular case, we may be confused. The "struct" in question, gerritclient.ProjectsFlag, is an alias for a map, not a struct.

Since it's a map, a native type, the encapsulation arguments I've made don't really apply, and it's more obvious to me to make it an empty map instead of a nil one. I'll make that change.

Thanks for the link! I'll keep this in mind.